### PR TITLE
Update to match standard Meteor login and Account token storage

### PR DESCRIPTION
**Pull request by jazeee**
Replaces prior pull request.

Meteor Accounts package uses hashedTokens instead of tokens. I updated the code to
match Accounts package.
Fix for https://github.com/kahmali/meteor-restivus/issues/79

### CODE REVIEW

**Comment by kahmali** [comment]
I'll review this later tonight, but it looks pretty darn good at first glance. There may be a few nit-picky things, but this should be close to being merged. Thanks for all your work in putting this together!

File: lib/restivus.coffee [hunk 1]
```diff
@@ -309,11 +311,19 @@ class @Restivus
       get: ->
         # Remove the given auth token from the user's account
         authToken = @request.headers['x-auth-token']
-        Meteor.users.update @user._id, {$pull: {'services.resume.loginTokens': {token: authToken}}}
+        hashedToken = Accounts._hashLoginToken authToken
+        tokenPaths = self.config.auth.token.split '.'
+        tokenPrefix = tokenPaths[..-2].join '.'
+        tokenSuffix = tokenPaths[-1..][0]
+        pullQueryTarget = {}
```
Comment (jazeee) [inline lib/restivus.coffee]: I wasn't 100% excited about this area of the code.
I didn't find a good way to put the query together as needed.

**Comment by kahmali** [comment]
Sorry, I didn't have time to review this last night. I'll be sure to go over it tonight. I'll keep you posted.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,22 @@
...
+## [Unreleased]
+
+#### Warning - Potentially breaking change
+- Restivus used to store the account login token in the user document: services.resume.loginTokens.token
+- Restivus now stores the account login token as a hashed token, in the user document: services.resume.loginTokens.hashedToken
+- This matches Meteor Accounts package
+
+#### Fixed
+- Issue #79:
+  - Update to match standard Meteor login and Account token storage
```
Comment (kahmali) [inline CHANGELOG.md]: This should be moved to the "Changed" section. Fixes are for bugs. This was an intended feature at one time that just needed to be updated to match the current API. I recommend you change it to:

- Update default auth endpoints to match current Accounts token storage (see #79)

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,22 @@
...
+
+#### Changed
+- Return "Unauthorized" for failed authentication
+- To match Meteor, storepassword token as hashedToken
```
Comment (kahmali) [inline CHANGELOG.md]: Just a small typo here: "storepassword" should be "store password". 

Also, it is nice if you always highlight code in markdown, so hashedToken should be `hashedToken`, just so users know it's something they can actually expect to see in the code.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,22 @@
 # Change Log
 
+## [Unreleased]
+
+#### Warning - Potentially breaking change
+- Restivus used to store the account login token in the user document: services.resume.loginTokens.token
+- Restivus now stores the account login token as a hashed token, in the user document: services.resume.loginTokens.hashedToken
+- This matches Meteor Accounts package
```
Comment (kahmali) [inline CHANGELOG.md]: Please highlight any code you use with markdown. I left a similar comment below. Helps me sleep better at night. So services.resume.loginTokens.token should be `services.resume.loginTokens.token`

File: README.md [hunk 5]
```diff
@@ -263,8 +263,8 @@ but all properties are optional):
 ##### `auth`
 - _Object_
   - `token`  _String_
-    - Default: `'services.resume.loginTokens.token'`
-    - The path to the auth token in the `Meteor.user` document. This location will be checked for a
+    - Default: `'services.resume.loginTokens.hashedToken'`
+    - The path to the auth hashed token in the `Meteor.user` document. This location will be checked for a
```
Comment (kahmali) [inline README.md]: Sounds better as "the path to the hashed auth token"

File: README.md [hunk 6]
```diff
@@ -286,7 +286,8 @@ but all properties are optional):
       - Partial auth
         - `userId`: The ID of the user being authenticated
         - `token`: The auth token to be verified
-        - If both a `userId` and `token` are returned, authentication will succeed if the `token`
+        - If both a `userId` and `token` are returned, Restivus will hash the token, then,
+          authentication will succeed if the `hashed token`
```
Comment (kahmali) [inline README.md]: Should be `hashedToken` (no space, and camel case) :)

File: lib/restivus.coffee (hunk 1 above)
Comment (kahmali) [inline lib/restivus.coffee]: I'm sure there are some small things that could be done to clean it up in the future, but for now it works, and it looks OK to me. I'd recommend a few variable name changes, for a bit more clarity:

`tokenPaths` -> `tokenLocation`
`tokenPrefix` -> `pathToTokens`
`tokenSuffix` -> `tokenFieldName`
`pullQueryTarget` -> `tokenToRemove`
`pullQuery` -> `tokenRemovalQuery`

I'm not in love with the last one, but it's more descriptive than `pullQuery`. If you think of anything better for any of them feel free to use that instead.

File: lib/route.coffee [hunk 7]
```diff
@@ -159,11 +159,12 @@ class @Route
     auth = @api.config.auth.user.call(endpointContext)
 
     # Get the user from the database
-    if not auth?.user and auth?.userId and auth?.token
-      userSelector = {}
-      userSelector._id = auth.userId
-      userSelector[@api.config.auth.token] = auth.token
-      auth.user = Meteor.users.findOne userSelector
+    if not auth?.user
+      if auth?.userId and auth?.token
```
Comment (kahmali) [inline lib/route.coffee]: Why the extra layer of nesting? This would short circuit if any check returns `false`. No real benefit, since we have to check for the `auth.user` again anyway.

**Comment by kahmali** [comment]
I left some comments for some minor edits, but aside from that, everything looks good to me. Of course, it wouldn't be a Restivus PR without me picking on someone's commit message, and of course yours doesn't meet my [strict standards](https://github.com/kahmali/meteor-restivus/blob/devel/CONTRIBUTING.md#committing) :D But instead of just leaving you to struggle with it, for now I just went ahead and wrote one for you that you can just drop right in: 

```
Use current Accounts token location in auth endpoints
    
- Update the default auth endpoints to store and remove tokens from the 
  location expected by the Meteor Accounts package 
  - The `accounts-base` package now stores hashed tokens in 
    `services.resume.loginTokens.hashedToken`, instead of storing a 
    `services.resume.loginTokens.token` unhashed. 
- Resolve #79
```

One cool thing about GitHub (since you said you were fairly new to it) is that it will automatically link to issues and PRs if you just use the shorthand `#<isssue or PR number>`. So #82 will link back to this PR, and It will even leave a note in the Issue or PR about where it is being referenced from.

**Comment by jazeee** [comment]
Sounds good. I will update to match.

File: lib/restivus.coffee (hunk 1 above)
Comment (jazeee) [inline lib/restivus.coffee]: I decided that the original code was too complicated. I switched to just finding the lastIndexOf '.'

It reduces the code by a line, and eliminates a variable. I'll commit it soon.

File: lib/route.coffee (hunk 7 above)
Comment (jazeee) [inline lib/route.coffee]: The original code was a little hard to parse as a human. In general, I prefer to explicitly define the logic in order to ensure all humans understand it.
In other words, a human could read the original code as:

    not ( user and userId and token )

When the CPU would read it as (I believe):

    ( not user ) and userId and token

This change is purely to make it easier for a human to understand the desired logic without question.

Comment (kahmali) [inline lib/route.coffee]: That's certainly a fair argument. For me, the extra layer of nesting makes me look for some further meaning, which slows me down a tiny bit. I wonder if a good compromise would be to rephrase it as 
```coffeescript
if auth?.userId and auth?.token and not auth?.user
```

What do you think of that? We get to keep the one-liner, and the ambiguity is removed. 

In reality, the `not` clause here isn't totally necessary. It only catches the weird situation where someone returned a `userId`, `token`, and `user` from their custom auth function (and skips the extra layer of auth if the `user` is included), when in reality they should only be returning either a `user` by itself or a `userId` and `token` together. After further review, I don't think the `userId` should be required either - but that will need to be taken care of in a separate PR.

File: lib/restivus.coffee [hunk 8]
```diff
@@ -309,11 +311,20 @@ class @Restivus
       get: ->
         # Remove the given auth token from the user's account
         authToken = @request.headers['x-auth-token']
-        Meteor.users.update @user._id, {$pull: {'services.resume.loginTokens': {token: authToken}}}
+        hashedToken = Accounts._hashLoginToken authToken
+        tokenLocation = self.config.auth.token
+        index = tokenLocation.lastIndexOf '.'
+        tokenPath = tokenLocation.substring 0, index
+        tokenFieldName = tokenLocation.substring index + 1
```
Comment (jazeee) [inline lib/restivus.coffee]: This is more straightforward logic with better variable names.

File: lib/route.coffee (hunk 7 above)
Comment (kahmali) [inline lib/route.coffee]: If you agree, feel free to remove the `not` clause entirely. I checked, and doing so would not conflict with the existing documentation.

Comment (jazeee) [inline lib/route.coffee]: That's fine. I updated it.
I am not fond of long if statements, since they suggest that we are doing too many things at once, and can also result in missed comprehension.
In any case, it works for me for now.

**Comment by jazeee** [comment]
Ok, I believe I have addressed all the comments.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,19 @@
 # Change Log
 
+## [Unreleased]
+
+#### Warning - Potentially breaking change
+- Restivus used to store the account login token in the user document: `services.resume.loginTokens.token`
```
Comment (kahmali) [inline CHANGELOG.md]: I think the first line in this warning should be something along the lines of: `All clients consuming a Restivus API will need to reauthenticate after this update, as their existing credentials will be rendered invalid".

**Comment by kahmali** [comment]
I know this is slightly evil of me (since it will require some changes to the docs and a slight code change), but I'm actually leaning towards going back to the original plan of making this change backwards-compatible. It's too jarring a change to not be considered API-breaking. Forcing all your clients to reauthenticate can cause users to feel like your application is broken when they didn't log out. Especially since these tokens didn't have expirations on them and may have been in use for a while. My only two choices at this point are to release this as a major update (0.7.0), or make it backwards compatible and move the API-breaking changes into 0.7.0. I think the latter is the better option, since this change really isn't major enough to require it's own major release. 

To make this backwards-compatible, we just need to switch `services.resume.loginTokens.token` back to the default token location in the API config. This way you can configure the proper location, and anyone else that needs it can do the same. I'll make the `hashedToken` the default location in the 0.7.0 update, since it has a ton of other API-breaking changes in it. If you need me to make the updates on your repo just let me know. I don't mind helping to make these changes, since I just sprung them up on you. Sorry about that!

**Comment by kahmali** [comment]
I'm not finished with the code review yet, but I realized that the bigger issue for you is storing and retrieving the hashed token, so making the smaller, non-API-breaking changes wouldn't help you much. I guess that's a good argument for just making this the 0.7.0 release and renaming the `0.7.0` branch to `0.8.0`. I'm now leaning towards that. Sorry for my indecisiveness. I just want to make the necessary changes in the most user-friendly way.

File: CHANGELOG.md (hunk 9 above)
Comment (jazeee) [inline CHANGELOG.md]: No problem. I updated it with modification.  (Their token is invalid, but their credentials are safe)

**Comment by jazeee** [comment]
Sounds good. I tend to agree that this is a 0.7.0 release.

File: CHANGELOG.md (hunk 9 above)
Comment (kahmali) [inline CHANGELOG.md]: Sorry for my incorrect use of terminology. You're right, it's their token, not their credentials, that will be invalid. Thanks for catching that.

File: lib/restivus.coffee (hunk 8 above)
Comment (kahmali) [inline lib/restivus.coffee]: I like this way of doing it as well. Definitely more straightforward, as you suggested.

**Comment by kahmali** [comment]
Okay, this looks great to me. You can go ahead and squash these all down into a single commit with the following message (updated for accuracy):

```
Use hashed token in authentication

- Validate against a hashed version of any provided token during auth 
  (previously validated against token directly)
- Update the default auth endpoints to store and remove tokens from the 
  location expected by the Meteor Accounts package 
  - The `accounts-base` package now stores hashed tokens in 
    `services.resume.loginTokens.hashedToken`, instead of storing a 
    `services.resume.loginTokens.token` unhashed. 
- Resolve #79
```

Once you do that just let me know and I'll merge these in. I've decided to release this as v0.7.0. I'm going to make an update to the README with a warning about updating to v0.7.0, then I'll publish it later tonight or early tomorrow. Thanks for the great contribution!

**Comment by jazeee** [comment]
I am having some trouble with rebasing. I think that the commit tree is funky, and I cannot figure out how to make rebase work as expected.

**Comment by kahmali** [comment]
I just looked at your commit history. Ouch. I'll see if I can rebase locally and I'll keep you posted with any progress I make. I pride myself on my super clean [commit history](https://github.com/kahmali/meteor-restivus/network), so we'll certainly need to clean this up before merging it in. These changes need to be reduced to a single commit, by any means necessary. :) It looks like you have some merges in there. You should never be merging in this project -- always rebase. Just a heads up. Especially when moving between branches for a single set of changes. I'm sure we can figure out how to set this right. 

In the meantime, here's a pretty solid guide on merging vs. rebasing in git: https://www.atlassian.com/git/tutorials/merging-vs-rebasing/

**Comment by jazeee** [comment]
I was trying to do the rebase process, with pick followed by squashes, but when I push back to github, it said I need to merge.
From what I read, you aren't supposed to rebase once you have published your repository. I am really in unfamiliar territory, so I am unsure of what it is supposed to do.

**Comment by kahmali** [comment]
Oh. That's an easy one, if I'm understanding you correctly. Sounds like you're doing the right thing. Once you rebase you just need to do a force push. When you rebase, you're rewriting your commit history, meaning the same changes will have a new SHA applied to them. So your local and remote branches are now out of sync (just do a `git log` and compare to the SHA on the remote branch your rebasing). If you're confident that the commits have been properly squashed together and your commit message is all set, you can just run
```bash
git push origin <feature-branch-name> --force
```
which will force the remote branch to match your local branch. 

Beware, doing a force push will overwrite your remote repo, and it's a fairly difficult thing to come back from (in fact, I'm not sure you can come back from it at all, although I have a clone of your repo so your okay for now -- the beauty of git is that it's not centralized, so each user has a completely history locally).

The problem with rebasing that you're referring to is with rebasing _public_ branches. Public branches are those where you reasonably expect someone to have their own history based off that branch. On Restivus, that's the `master`, `devel`, and `0.7.0` branches. I can and have rebased any other branches, even remote ones, as long as I'm the only one working from them, or I can coordinate closely with anyone else that is. If you check out the link I provided, they list it as the "Golden Rule of Rebasing," and it absolutely is. I've screwed it up once on Restivus, and hopefully it's the last time I ever do it. :)

**Comment by jazeee** [comment]
Ahh, makes sense. I figured it was an issue like that.
I will try that.
Thanks.

**Comment by jazeee** [comment]
Ok, that fixed it.
I cleaned it up, so it should have only one commit now, with all the changes on branch fix-issue-79
Thanks for the help.

**Comment by kahmali** [comment]
OK, I gave it one final review and everything looks good on my end. I'm going to merge this in. I'll get the update published by tomorrow. I'll keep you posted. Thanks so much for all your work on this!

**Comment by jazeee** [comment]
Excellent. Thanks much for all the help too.

**Comment by kahmali** [comment]
No problem! Glad to help!

File: lib/route.coffee [hunk 10]
```diff
@@ -159,10 +159,10 @@ class @Route
     auth = @api.config.auth.user.call(endpointContext)
 
     # Get the user from the database
-    if not auth?.user and auth?.userId and auth?.token
+    if auth?.userId and auth?.token and not auth?.user
       userSelector = {}
       userSelector._id = auth.userId
-      userSelector[@api.config.auth.token] = auth.token
+      userSelector[@api.config.auth.token] = Accounts._hashLoginToken auth.token
```
Comment (kahmali) [inline lib/route.coffee]: Almost missed this:

Hashing the token should actually be done from the default auth function. This way it doesn't interfere with the custom auth discussed in #6, and it will also play nicely with #49. Otherwise, users with custom auth would be forced to hash their tokens the same way, which doesn't provide enough flexibility. I'll fix this myself, but I just wanted to run it by you in case you had any objections.

Comment (jazeee) [inline lib/route.coffee]: That sounds fine to me. Of course, as long as the unit tests pass! Lol
